### DISCUSSION TITLE
Intellisense for code blocks?

### DISCUSSION THREAD

**Discussion by lucastobrazil** [mdx-js/mdx#2429] [Ideas]
Would be cool to be able to get intellisense for code blocks too. Even though they may be rendered as strings in the output, whilst writing code blocks it would be very useful to know that the code blocks are using the right types (and eslinted too)

For context, we're using MDX for a component library documentation site (Build with Gatsby 5) with editable code snippets. See https://www.vitalitydesignsystem.com/components/avatar/ as an example,

In order to render the editable code (uses [React-Live](https://github.com/FormidableLabs/react-live) under the hood), we need to place all the code blocks in back ticks, so ideally Intellisense can also read within those code blocks too.

If i understand correctly, MDX renders the code blocks as a string, so perhaps intellisense cannot process them... But i'd hope for something like:

```mdx
import { MyTypedComponent } from '../components/MyTypedComponent.js'

# My MDX File

<MyTypedComponent foo="bar" /> // intellisense works

.```jsx
<MyTypedComponent foo="bar" /> // intellisense doesn't work but should
                       ~~~~ // squiggly lines from eslint would also be helpful
.```

**Comment by remcohaszing** [comment]
I think you means this?

````mdx
import { MyTypedComponent } from '../components/MyTypedComponent.js'

# My MDX File

<MyTypedComponent foo="bar" /> // intellisense works

```jsx
<MyTypedComponent foo="bar" /> // intellisense doesn't work but should
                       ~~~~ // squiggly lines from eslint would also be helpful
```
````

I have fiddled with that, but it’s a complex feature. Support tracked in https://github.com/mdx-js/mdx-analyzer/issues/287.

**Reply by wooorm** [reply]
Not sure it should work? Code blocks often include partial code that depends on each other. And includes broken code. And redefines things.

**Reply by remcohaszing** [reply]
We should determine per feature what is or isn’t allowed. If and how well each feature works really depends on the VSCode extensions installed, and what impact the feature has. For example support for completion and hover information would be great, but I wouldn’t want to show error/warning/info diagnotics. Those could be demoted to hints though.

**Reply by wooorm** [reply]
Completion and hover info could be wrong though? It assumes things are loaded and configured a certain way, that likely is not that way in the code that’s talked about?

**Reply by lucastobrazil** [reply]
> Not sure it should work? Code blocks often include partial code that depends on each other. And includes broken code. And redefines things.

Yes and no - I think it's fair to accept that sometimes people will write code blocks that are intended to be complete and correct. I don't see why code blocks should miss out on intellisense just because people might write pseudocode etc. 

Maybe there's a middle ground?

**Reply by remcohaszing** [reply]
The experience of editing code blocks will be similar to when you create a new file in VSCode in memory. Also this is both a low prio and a complex issue pushing the boundaries of what Volar can do. Even then it requires actual editor support, a language server isn’t sufficient. So this feature would only work for VSCode.

I want to implement this at some point, but don’t this anytime soon.
